piper_backend.py

```diff
@@ -0,0 +1,215 @@
+
+logger = logging.getLogger(__name__)
+
+
+class Piper(TTS):
+    def __init__(self, parsed):
+        super().__init__()
+        self.parsed = parsed
+        self.paused = False
+        self.reset_issued = Locked(False)
+        self.play_queue = queue.Queue()
+        self.gen_queue = queue.Queue()
+        self.get_queue = queue.Queue()
+        self.play_queue_size = Locked(0)
+        self.gen_queue_size = Locked(0)
+        self.gen_process = Locked(None)
+        self.play_process = Locked(None)
+
+        self.ffplay_path = shutil.which("ffplay")
+
+        self.piper_path = shutil.which("piper-tts")
+        if self.piper_path is None and not self.parsed.piper_python:
+            logger.warning("The piper C++ executable was not found")
+
+        self.is_piper_python = self.piper_path is None or self.parsed.piper_python
+        if self.is_piper_python:
+            if importlib.util.find_spec("piper") is None:
+                logger.critical("The piper python module was not found")
+                self.inited = False
+                return
+
+        self.gen_thread = threading.Thread(target=self.run_gen_thread, daemon=True)
+        self.play_thread = threading.Thread(target=self.run_play_thread, daemon=True)
+
+        self.gen_thread.start()
+        self.play_thread.start()
+        self.inited = True
+
+    def run_play_thread(self):
+        while True:
+            audio = self.play_queue.get()
+            if self.reset_issued.get():
+                self.play_queue.task_done()
+                continue
+
+            try:
+                self.play_process.set(
+                    subprocess.Popen(
+                        [
+                            self.ffplay_path,
+                            "-hide_banner",
+                            "-loglevel",
+                            "panic",
+                            "-nostats",
+                            "-autoexit",
+                            "-nodisp",
+                            "-af",
+                            f"atempo={self.parsed.speed},volume={self.parsed.volume}",
+                            "-f",
+                            "s16le",
+                            "-ar",
+                            f"{self.parsed.piper_rate}",
+                            "-ac",
+                            "1",
+                            "-",
+                        ],
+                        stdin=subprocess.PIPE,
+                    )
+                )
+                self.play_process.get().communicate(input=audio)
+            finally:
+                self.play_process.set(None)
+                self.play_queue.task_done()
+                with self.play_queue_size.lock:
+                    self.play_queue_size.data = self.play_queue_size.data - len(audio)
+                    self.play_queue_size.data = max(0, self.play_queue_size.data)
+
+    def run_gen_thread(self):
+        while True:
+            text, getaudio = self.gen_queue.get()
```

Comment:
We should probably add back the text sanitization, which is really useful in some cases, especially for some PDFs
```py
unidecode(text).replace("‐\n", "").replace("‐ ", "")
```

Reply:
I forgot about that. Done in main.py:read()